MoreLinq.NoConflictGenerator/Program.cs

```diff
@@ -0,0 +1,227 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#endregion
+
+namespace MoreLinq.NoConflictGenerator
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Reflection;
+    using System.Text.RegularExpressions;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+    static class Program
+    {
+        static void Run(IEnumerable<string> args)
+        {
+            var dir = Directory.GetCurrentDirectory();
+
+            string includePattern = null;
+            string excludePattern = null;
+            var usings = new List<string>();
+
+            Exception MissingArgValue() =>
+                new InvalidOperationException("Missing argument value.");
+
+            using (var arg = args.GetEnumerator())
+            {
+                while (arg.MoveNext())
+                {
+                    switch (arg.Current)
+                    {
+                        case "-i":
+                        case "--include":
+                            includePattern = Read(arg, MissingArgValue);
+                            break;
+                        case "-x":
+                        case "--exclude":
+                            excludePattern = Read(arg, MissingArgValue);
+                            break;
+                        case "-u":
+                        case "--using":
+                            usings.Add(Read(arg, MissingArgValue));
+                            break;
+                        case "":
+                            continue;
+                        default:
+                            dir = arg.Current[0] != '-'
+                                ? arg.Current
+                                : throw new Exception("Invalid argument: " + arg.Current);
+                            break;
+                    }
+                }
+            }
+
+            var includePredicate
+                = string.IsNullOrEmpty(includePattern)
+                ? (_ => true)
+                : new Func<string, bool>(new Regex(includePattern).IsMatch);
+
+            var excludePredicate
+                = string.IsNullOrEmpty(excludePattern)
+                ? (_ => false)
+                : new Func<string, bool>(new Regex(excludePattern).IsMatch);
+
+            var thisAssemblyName = typeof(Program).GetTypeInfo().Assembly.GetName();
+
+            var q =
+                from fp in Directory.EnumerateFiles(dir, "*.cs")
+                select new
+                {
+                    SourcePath = fp,
+                    Methods =
+                        from cd in
```

Comment:
Hmm…I could throw it into a local function but not sure how much it'll help (see below). It might un-nest things but it reads disjointed. Moving to a separate method means the code is even further away from its sole usage.

```c#
            IEnumerable<MethodDeclarationSyntax> GetMethods(string source) =>
                from cd in
                    CSharpSyntaxTree
                        .ParseText(source, CSharpParseOptions.Default.WithPreprocessorSymbols("MORELINQ"))
                        .GetRoot()
                        .SyntaxTree
                        .GetCompilationUnitRoot()
                        .DescendantNodes().OfType<ClassDeclarationSyntax>()
                where (string) cd.Identifier.Value == "MoreEnumerable"
                from md in cd.DescendantNodes().OfType<MethodDeclarationSyntax>()
                let mn = (string) md.Identifier.Value
                where md.ParameterList.Parameters.Count > 0
                && md.ParameterList.Parameters.First().Modifiers.Any(m => (string)m.Value == "this")
                && md.Modifiers.Any(m => (string)m.Value == "public")
                && md.AttributeLists.SelectMany(al => al.Attributes).All(a => a.Name.ToString() != "Obsolete")
                && includePredicate(mn)
                && !excludePredicate(mn)
                select md;

            var q =
                from fp in Directory.EnumerateFiles(dir, "*.cs")
                select new
                {
                    SourcePath = fp,
                    Methods = GetMethods(File.ReadAllText(fp))
                };
```